### PR TITLE
Add ability to mix usage of T2I-Adapter(s) and ControlNet(s).

### PR DESCRIPTION
**Pull request by GreggHelt2**
# What does this PR do?

Adds ability to mix usage of T2I-Adapter(s) and ControlNet(s). Previously, UNet2DConditional implementation only allowed use of one or the other. Adds new forward() arg `down_intrablock_additional_residuals` specifically for T2I-Adapters. If `down_intrablock_addtional_residuals` is not used, maintains backward compatibility with prior usage of only T2I-Adapter or ControlNet but not both, via `down_block_additional_residuals` and `mid_block_additional_residual args`.

Here's an example of mixing T2I-Adapter and ControNets with this fix, (in an InvokeAI dev branch). From left to right: original image, then inference with prompt "old man" and same seed -- ControlNet Canny, T2I-Adapter Color, both applied together.
Screenshot from 2023-10-10 08-02-10

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [X] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
The problem with combined use of T2I-Adapter and ControlNet was mentioned during [initial discussions of diffusers T2I-Adapter implementation](https://github.com/huggingface/diffusers/pull/2555#issuecomment-1473248404)
- [X] Did you make sure to update the documentation with your changes? 
I added documentation for UNet2DConditional.forward() existing args
`down_block_additional_residuals` and `mid_block_additional_residual` and new arg `down_intrablock_additional_residuals`
- [ ] Did you write any new necessary tests?

## Who can review?
@patrickvonplaten
@sayakpaul
@williamberman

### CODE REVIEW

File: src/diffusers/models/unet_2d_condition.py
Comment (psychedelicious) [inline src/diffusers/models/unet_2d_condition.py:812]:
```suggestion
            down_block_additional_residuals (`tuple` of `torch.Tensor`, *optional*):
                additional residuals to be added to UNet long skip connections from down blocks to up blocks
                for example from ControlNet side model(s)
            mid_block_additional_residual (`torch.Tensor`, *optional*):
                additional residual to be added to UNet mid block output, for example from ControlNet side model
            down_intrablock_additional_residuals (`tuple` of `torch.Tensor`, *optional*):
                additional residuals to be added within UNet down blocks, for example from T2I-Adapter side model(s)
```

Comment (GreggHelt2) [reply src/diffusers/models/unet_2d_condition.py]: Thanks for taking a look, added your suggestions.

File: src/diffusers/models/unet_2d_condition.py
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:781]: Ok for me! Think it's indeed a good idea to cleanly separate ControlNet and T2I here

File: src/diffusers/models/unet_2d_condition.py
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:997]:
```suggestion
        if not is_adapter and mid_block_additional_residual is None and down_block_additional_residuals is not None:
            deprecate("T2I should not use down_block_additional_residuals", "1.3.0", "Passing intrablock residual connections with `down_block_additional_residuals` is deprecated in will be removed in diffusers 1.3.0.  `down_block_additional_residuals` should only be used for ControlNet. Please make sure use `down_intrablock_additional_residuals` instead. ", standard_warn=False) <- let's deprecate using "down_block_additional_residuals" for T2I adapters
            down_intrablock_additional_residuals = down_block_additional_residuals
            is_adapter = True
```

**Review by patrickvonplaten — COMMENTED** [review]
The change makes sense to me - we should try to keep UNet2D as general as possible to allow for such use cases. 

Before merging it would be good to:
- 1.) Adapt all T2I pipelines to pass `down_intrablock_additional_residuals` instead of `down_block_additional_residuals` 
- 2.) Add a test that makes sure that the legacy format is still supported.

Also I'd very much like to get a review from @williamberman here

**Review by williamberman — APPROVED** [review]
Love this, `down_intrablock_additional_residuals` is a bit of a wonky name so if there is a better name we can come up with (I can't think of one), that would be great but if not lgtm

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Tests are passing, PR is a nice refactor that disentangles two concepts that used the same function input. Will fix quality test directly on main.

**Pull request merged**
